### PR TITLE
Candidates can withdraw applications

### PR DESCRIPTION
**Pull request by duncanjbrown**
### Context

Candidates need to be able to withdraw application choices via the UI. This PR adds that feature

#### What it does

- changes the state of the application to `withdrawn` and sets a `withdrawn_at` date in the DB
- notifies Slack
- tries to calculate the DBD date

#### What it doesn't do

- email anyone
- implement any designs that have been Actually Designed by Designers

### Changes proposed in this pull request

🔧 New `WithdrawApplication` service
⛓ Slack integration for the service
🕐 DBD integration for the service
✋ Rudimentary UI
🇬🇧 Feature flag

Plus a bit of refactoring of specs.

### Withdrawal link

Screenshot 2019-12-04 at 12 37 57

### Confirming withdrawal

Screenshot 2019-12-04 at 12 38 01

### It's withdrawn

Screenshot 2019-12-04 at 12 38 09

### Guidance to review

The feature is controlled by a feature flag which you will need to turn on in order to test it through the front end. 

There is no spec which looks at the Decline by Default behaviour of this service end-to-end but I feel confident we have enough specs so far — does the level of testing feel right?

### Link to Trello card

https://trello.com/c/OA9ZrEld/1339-candidates-can-withdraw-applications-%F0%9F%8F%88

### Env vars

N/A

### CODE REVIEW

File: app/lib/state_change_notifier.rb [hunk 1]
```diff
@@ -32,6 +33,8 @@ def self.call(event, candidate: nil, application_choice: nil, application_form:
       url = helpers.support_interface_application_form_url(application_form_id)
     when :offer_declined
       text = ":no_good: #{applicant} has declined #{provider_name}'s offer"
+    when :withdraw
+      text = "#{applicant} has withdrawn their application for #{course_name} at #{provider_name}"
```
Comment (tijmenb) [inline app/lib/state_change_notifier.rb]: Needs emoji!

File: app/services/withdraw_application.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class WithdrawApplication
+  def initialize(application_choice:)
+    @application_choice = application_choice
+  end
+
+  def save!
+    ActiveRecord::Base.transaction do
+      ApplicationStateChange.new(application_choice).withdraw!
+      application_choice.update(withdrawn_at: Time.zone.now)
```
Comment (tijmenb) [inline app/services/withdraw_application.rb]: If we're not checking the return value:

```suggestion
      application_choice.update!(withdrawn_at: Time.zone.now)
```

**Comment by duncanjbrown** [comment]
Thanks @tijmenb, will rebase and fix up

**Comment by duncanjbrown** [comment]
All should be resolved ✅

**Comment by paulrobertlloyd** [comment]
@duncanjbrown Small thing; we should have a red button for ‘Withdraw my application’, and the withdrawn status badge should also be red.

**Comment by chubberlisk** [comment]
> @duncanjbrown Small thing; we should have a red button for ‘Withdraw my application’, and the withdrawn status badge should also be red.

For updating the withdrawn status badge to red, have a look at https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/626 for how the other statuses have been done. (Hopefully saves you a few moments to work how we do this 😉)

**Comment by duncanjbrown** [comment]
🙏🙏🙏 Thank you @paulrobertlloyd and @chubberlisk

**Comment by duncanjbrown** [comment]
Done

**Pull request merged**